### PR TITLE
[Discussion] `ConfigurableObject`

### PR DESCRIPTION
**Pull request by hiqsol**
How about adding `ConfigurableObject` class implementing `configure($config)` in constructor.

We still want most of the classes to have normal constructors and don't be configurable.
But there are classes where it is quite suitable.

### CODE REVIEW

**Comment by samdark** [comment]
How about `Configurable` trait instead?

**Comment by rob006** [comment]
Why would you need a base object (or even trait) for such thing? Calling methods in this way is strange (and IMO unexpected), and special class for 3 lines of code is an overkill.

**Comment by samdark** [comment]
What are use-case classes additional to Model instances?

File: src/base/ConfigurableObject.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * {
+ *     ...
+ *     parent::__construct($config);
+ * }
+ * ```
+ *
+ * That is, a `$config` parameter (defaults to `[]`) should be declared as the last parameter
+ * of the constructor, and the parent implementation should be called at the end of the constructor.
+ */
+class BaseObject
```
Comment (FabrizioCaldarelli) [inline src/base/ConfigurableObject.php]: Should not class name be "ConfigurableObject" instead Object ?

File: src/base/ConfigurableObject.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+{
+    public function __construct(iterable $config = [])
+    {
+        if (!empty($config)) {
```
Comment (FabrizioCaldarelli) [inline src/base/ConfigurableObject.php]: Why don't we call `AbstractContainer::configure($this, $config)` ?

And remove static function configure.

**Comment by hiqsol** [comment]
@rob006 yes it saves not that many lines of code. But it saves discussions and explanations.
We say or see `extends ConfigurableObjects` and we know what is it all about.

**Comment by rob006** [comment]
@hiqsol For me `array $config = []` in constructor is enough, no need to create redundant type for this. If we want to make it simpler, static helper like `Yii::configure()` would be much more flexible. 

I'm also strongly against allowing calling methods via `$config` - it is unexpected and very likely will be source of security issues (especially with current implementation, which allows accessing protected properties and methods).

**Comment by hiqsol** [comment]
@samdark, @SilverFire, @cebe, @machour 
Let's vote and make a decision.

**Comment by samdark** [comment]
I am against having such generalized base class. An interface and a trait are OK though.

**Comment by machour** [comment]
+1 for introducing something like this, adapted as a trait as proposed by @samdark, which would allow making a single class configurable in my class hierarchy.

**Comment by FabrizioCaldarelli** [comment]
> I am against having such generalized base class. An interface and a trait are OK though.

Anyway, an interface or trait will require the class to be created, using direct "new" or static method.

Such as:

```
trait ConfigurableTrait
{
      public function fromConfig($config = [])
      {
              \yii\di\AbstractContainer::configure($this, $config);
      }
}

class MyClass
{
     use ConfigurableTrait;
}
```

`
$model = (new MyClass())->fromConfig($config);
`

**Comment by cebe** [comment]
There are many ways to implement it which have pros and cons:

I think we should come from use cases to an implementation and not the other way round. The currently proposed implementation has several drawbacks:

- configure() is static member of the same class so it can set private and protected members
- having it as a base class does not allow adding it to existing classes which follow a differnet hierarchie. This was no problem in Yii 2 where all classes extended from it anyway, but makes it hard to use in Yii 3 when you can not easily add it when needed.

The alternative of using a trait does not allow for checking it, e.g. we can not test if `$object instanceof Configurable` which could be useful. Also that trait would only include a simple constructor which one could also write out instead of hiding the functionality in a trait.

So we can provide the concept of configurable objects in Yii 3 by allowing constructors like `__construct(..., array $config)` and make the DI part aware of how to pass a config to that.
Also provide the static `configure()` method to be used inside of such constructors. No trait or base class is needed.

**Comment by miolae** [comment]
May be there is a reason to take "ConfigurableObject" out of the main part to a component similar to [symfony/config](https://github.com/symfony/config)? Or maybe we could use symfony component itself (but it is much more complicated then just arrays).

**Comment by samdark** [comment]
symfony/config is a different thing.

**Comment by samdark** [comment]
Overall I think that is a wrong way to go. Objects should not care much how they're configured (except edge cases). DI container should take care of it. That way we'll be able to configure anything.

**Pull request closed without merging**